File: frontend/src/utils/geo.ts
/** Haversine distance between two WGS84 points, in kilometres. */
export function haversineKm(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

/** Web Mercator (EPSG:3857) half-circumference in metres: π × 6 378 137 */
const WEB_MERCATOR_EXTENT = 20037508.34;

/** Convert WGS84 lat/lon to SVG-coordinate Web Mercator (y-down). */
export function toSvgMercator(lon: number, lat: number): [number, number] {
  const x = (lon * WEB_MERCATOR_EXTENT) / 180;
  const latRad = (lat * Math.PI) / 180;
  const y =
    -(Math.log(Math.tan(Math.PI / 4 + latRad / 2)) * WEB_MERCATOR_EXTENT) /
    Math.PI;
  return [x, y];
}

File: backend/app/logic/spatial/geo.py
"""Scalar geodesic utilities."""

from math import atan2, cos, radians, sin, sqrt

EARTH_RADIUS_KM = 6371.0

type Coords = list[tuple[float, float]]  # [lon, lat] GeoJSON order


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Great-circle distance in km between two (lat, lon) points."""
    dlat = radians(lat2 - lat1)
    dlon = radians(lon2 - lon1)
    a = (
        sin(dlat / 2) ** 2
        + cos(radians(lat1)) * cos(radians(lat2)) * sin(dlon / 2) ** 2
    )
    return EARTH_RADIUS_KM * 2 * atan2(sqrt(a), sqrt(1 - a))


def total_length_km(coords: Coords) -> float:
    """Total length of a coordinate list in km. Coords are (lon, lat) GeoJSON order."""
    return sum(
        haversine_km(coords[i][1], coords[i][0], coords[i + 1][1], coords[i + 1][0])
        for i in range(len(coords) - 1)
    )
